### ISSUE TITLE
Cache-safe CellValues

### ISSUE THREAD

**Issue by fverdugo** [gridap/Gridap.jl#91]
The concrete `IndexCellValue` types that use a `CachedArray` internally **are not cache-safe!** When you get type `a[i]` you get a reference to the internal `CachedArray`. Further calls  `a[j]` modify the `CachedArray` and thus also the result you extracted with `a[i]` previously. 

An example:

```julia
julia> using Gridap

julia> using Gridap.CellValuesGallery

julia> a = CellValueFromArray([[1,2],[3,4]])
1 -> [1, 2]
2 -> [3, 4]


julia> b = a + a
1 -> [2, 4]
2 -> [6, 8]


julia> b1 = b[1]
2-element Gridap.CachedArrays.CachedArray{Int64,1,Array{Int64,1}}:
 2
 4

julia> b2 = b[2]
2-element Gridap.CachedArrays.CachedArray{Int64,1,Array{Int64,1}}:
 6
 8

julia> b1
2-element Gridap.CachedArrays.CachedArray{Int64,1,Array{Int64,1}}:
 6
 8

```

This is not a problem in our code since we never visit two elements at the same time. However, this makes the code thread-unsafe, when considering thread-based parallelism! 

**To think a thread-safe design of the `CellValue` machinery**

**Comment by santiagobadia** [comment]
It is not that simple
https://github.com/gridap/Gridap.jl/blob/0a72671f519e49a411b75d27c42da84f6e1d8627/src/Geometry/Simplexify.jl#L38-L46
the graph and face label generation does not rely on the grid.
Only orienting the grid I have ended up with non-consistent models. I have to deep into these methods to see how I have to modify them accordingly.

**Comment by fverdugo** [comment]
This has been addressed in v0.7.0. Closing.
